waspc/data/Generator/templates/server/src/auth/providers/oauth/state.ts

```diff
@@ -64,54 +68,64 @@ export function validateAndGetOAuthState<OT extends OAuthType>({
   provider: ProviderConfig,
   req: ExpressRequest
 }): OAuthStateWithCodeFor<OT> {
-  return {
-    ...validateAndGetCode(req),
-    ...validateAndGetState(provider, req),
-    ...(oAuthType === 'OAuth2WithPKCE' && validateAndGetCodeVerifier(provider, req)),
+  const state: OAuthStateWithCodeFor<OT> = {
+    ...getCode(req),
+    ...getState(req),
+    ...(oAuthType === 'OAuth2WithPKCE' && getCodeVerifier(provider, req)),
   };
-}
 
-function generateAndStoreState(
-  provider: ProviderConfig,
-  res: ExpressResponse
-): { state: string } {
-  const state = generateState();
-  setOAuthCookieValue(provider, res, 'state', state);
+  validateOAuthState(provider, req, state);
 
-  return { state };
+  return state;
 }
 
-function generateAndStoreCodeVerifier(
+function storeOAuthStateInCookies<OT extends OAuthType>(
   provider: ProviderConfig,
-  res: ExpressResponse
-): { codeVerifier: string } {
-  const codeVerifier = generateCodeVerifier();
-  setOAuthCookieValue(provider, res, 'codeVerifier', codeVerifier);
-
-  return { codeVerifier };
+  res: ExpressResponse,
+  state: OAuthStateFor<OT>
+): void {
+  setOAuthCookieValue(provider, res, 'state', state.state);
+  if (isOAuthStateWithPKCE(state)) {
```

Comment:
Why is this typeguard necessary?

We'll have to add a new "save" statement whenever we add a field, and the type system does not enforce it (i.e., it's possible to forget).

Why not simply loop over the state object and store everything that's inside? That way, if we add a new field, it will work automatically.

Reply:
Because state is a union of `OAuthState` and `OAuthStateWithPKCE`. 

As I mentioned here: https://github.com/wasp-lang/wasp/pull/1993#discussion_r1647687897

Iterating over the values requires from me to assert the types 😢 I can do that ofc